### ISSUE TITLE
calibrate train (binary): default fit on 50 rows does not finish in hours; time is in the exact-joint κ outer Hessian's fourth-order row jets

### ISSUE THREAD

**Issue by SauersML** [SauersML/gnomon#2359]
## Summary

At 705af772 (gam 0.2.1), `train` of a binary phenotype with default options does not finish on tiny data. On 50 rows with 0 PCs it was still running after 6,700 s on 2 threads, when it was stopped. Every stack sample is in the κ exact-joint outer optimization, building the outer Hessian's IFT correction from per-row fourth-order contracted derivatives. The info log goes silent in that phase.

## Repro

Build: `cargo build --release --locked --no-default-features --features calibrate --bin gnomon-calibrate` (nightly-2026-09-10, `-C target-cpu=native`, Xeon Platinum 8259CL).

```
GNOMON_LOG_LEVEL=info RAYON_NUM_THREADS=2 taskset -c 0,1 ./gnomon-calibrate train train.tsv --num-pcs 0
```

`train.tsv` (50 rows, 21 cases) is attached below. Stdout reaches "Auto-detected link function: Logit", the log shows one warning ("latent z is not close to standard normal; using global empirical latent-z calibration internally"), and no `model.json` is written.

## Runtimes (all stopped by hand, none finished)

| Rows | PCs | Options | Threads | Wall when stopped |
|---|---|---|---|---|
| 50 | 0 | defaults | 2 | 6,700 s |
| 100 | 0 | defaults | 2 | 6,700 s |
| 200 | 0 | defaults | 4 | 15,513 s |
| 200 | 2 | defaults | 4 | 15,513 s |
| 2,000 | 0 | defaults | 2 | 35,348 s |
| 2,000 | 2 | defaults | 4 | about 34,000 s |
| 2,000 | 0 and 2 | `--reml-max-iterations 5 --reml-convergence-tolerance 1e-2` | 2 | about 20,900 s |

Even with 5 outer iterations, the n = 2,000 fits ran past 5.8 h. Earlier, at n = 200 with 0 PCs, `--reml-max-iterations 1` on 1 thread ran past 600 s, and `--reml-max-iterations 0` panics (#2347).

## Where the time goes

`gdb -batch -ex "thread apply all bt"`: 5 samples on the 50-row run, 3 on each 200-row run. In every sample, the working threads sit on this path (outermost first):

```
gam::solver::workflow::fit_model
gam::families::bernoulli_marginal_slope::fit_bernoulli_marginal_slope_terms::{closure#11}
gam::terms::smooth::optimize_spatial_length_scale_exact_joint
<gam::solver::outer_strategy::OuterProblem>::run  /  run_outer_with_plan
gam::solver::estimate::reml::unified::reml_laml_evaluate
gam::solver::estimate::reml::unified::compute_outer_hessian
gam::solver::estimate::reml::unified::compute_ift_correction_trace
<OwnedJointDerivProvider as HessianDerivativeProvider>::hessian_derivative_correction_result
gam::families::custom_family::exact_newton_d2h_closure_owned
<BernoulliMarginalSlopeExactNewtonJointHessianWorkspace as ExactNewtonJointHessianWorkspace>::...
BernoulliMarginalSlopeFamily::row_primary_fourth_contracted_recompute(_ordered)
BernoulliMarginalSlopeFamily::empirical_flex_row_fourth_contracted_recompute
BernoulliMarginalSlopeFamily::empirical_flex_neglog_jet
BernoulliMarginalSlopeFamily::empirical_flex_eta_and_eta_a_jet_at_z
BernoulliMarginalSlopeFamily::for_each_deviation_basis_cubic_at
gam::families::jet_partitions::MultiDirJet::{mul, compose_unary}
```

An earlier sample of an n = 2,000 run was on the third-order sibling (`exact_newton_dh_closure_owned` → `row_primary_third_contracted_recompute`). Before the silent phase, the log's cache lines read `[BMS exact-cache] affine tail-cell memo: hits=0 misses=0 entries=0` and `[BMS cell-moment LRU] cycle hits=0 misses=0`. The intercept warm start reports `preseeded=0 (cold)`.

So the exact outer Hessian, over the log smoothing parameters and κ jointly, recomputes fourth-order multi-direction jets per row, per empirical latent node and per hyperparameter pair, at every outer evaluation. At 50 rows that cost doesn't end within hours.

<details><summary>train.tsv (50 rows)</summary>

```
sample_id	phenotype	score	sex
T0000000	1	0.7542257308959961	0
T0000001	1	0.22320079803466797	1
T0000002	0	-0.22734832763671875	0
T0000003	0	-0.7349481582641602	0
T0000004	1	-1.0814876556396484	1
T0000005	0	-0.3175668716430664	0
T0000006	1	1.4301624298095703	0
T0000007	0	-1.4387292861938477	1
T0000008	0	-1.1304388046264648	1
T0000009	0	-3.1554346084594727	0
T0000010	1	-0.058302879333496094	1
T0000011	0	-0.29225635528564453	0
T0000012	0	-0.24579906463623047	0
T0000013	0	0.11749553680419922	0
T0000014	1	0.3434133529663086	0
T0000015	1	-0.32691192626953125	0
T0000016	0	0.3752412796020508	1
T0000017	0	-0.05824565887451172	0
T0000018	0	-0.4093008041381836	0
T0000019	1	0.22971820831298828	0
T0000020	0	-0.4925365447998047	1
T0000021	0	0.36133480072021484	0
T0000022	1	-0.4608573913574219	1
T0000023	1	0.10126876831054688	1
T0000024	1	1.8222789764404297	0
T0000025	1	1.245779037475586	1
T0000026	1	0.916015625	0
T0000027	0	0.0823984146118164	0
T0000028	0	0.09211349487304688	1
T0000029	0	-0.4771699905395508	1
T0000030	0	-0.4767570495605469	0
T0000031	1	0.5295619964599609	0
T0000032	1	1.6902036666870117	1
T0000033	1	0.4617900848388672	0
T0000034	0	1.0950126647949219	0
T0000035	1	0.7325677871704102	0
T0000036	1	1.5006036758422852	0
T0000037	0	0.8378829956054688	1
T0000038	0	-1.2965764999389648	1
T0000039	0	-0.25040626525878906	0
T0000040	0	1.0696144104003906	0
T0000041	0	1.4821605682373047	1
T0000042	0	-1.525289535522461	0
T0000043	1	-1.4461479187011719	0
T0000044	0	-1.574213981628418	1
T0000045	1	0.5692768096923828	1
T0000046	0	0.39346885681152344	1
T0000047	0	1.234334945678711	0
T0000048	1	-0.04082012176513672	1
T0000049	0	0.4090251922607422	1
```

</details>

**Comment by SauersML** [comment]
The whole runtime is inside gam 0.2.1's exact-joint outer Hessian (every stack sample is in `compute_ift_correction_trace` recomputing fourth-order row jets), so this is a gam defect rather than something gnomon's request construction can change. Filed upstream as SauersML/gam#2922 with the runtimes and the stack. gnomon stays on 0.2.1 until a gam release carries the fix; this issue stays open to track the pin bump.

**Comment by SauersML** [comment]
## Status from gam: the hang is fixed on gam main, and the default train is still slow and stops early

**The hang.** The profile above is from gam 0.2.1: every sample is inside the exact-joint outer Hessian. gam main no longer builds that Hessian for this request.
- Since gam b7e2d19c61, "Skip joint-spatial outer when user fixes Duchon length_scale (margslope)": the explicit Duchon `length_scale` locks κ. The marginal-slope outer search is BFGS with no outer Hessian (`hessian=Unavailable` in the derivative-policy line).
- b7e2d19c61 is not in 0.2.1, and gnomon pins 0.2.1. So the fix reaches gnomon through the gam pin bump in gnomon-2370's lane (gnomon#2370).

**The default train on gam main** is gnomon's default binary calibration request, measured with test-profile gam builds on acl42:

| rows | gam SHA | wall | outer iterations | outcome |
|---|---|---|---|---|
| 50 | e4a2051e | 17 s | — | completes |
| 200 | e4a2051e | 371.75 s | 16 | completes |
| 200 | a35efd6588 | 485.14 s | 24 | completes |
| 2,000 | e4a2051e | 4,802.31 s | 22 | all log λ railed at 18-27, the fully smoothed corner |

The timings are sequential single runs, not interleaved A/B rounds.

**The default route also stops short of the converged reference.** Default tolerance against a tight reference, at gam a35efd6588:

| rows | criterion (default / tight) | max\|Δβ\| | max\|Δρ\| | railed coords (default / tight) | \|g\| (default / tight) |
|---|---|---|---|---|---|
| 50 | 35.58784 / 35.51656 | 1.36e-2 | 10.1 | [] / [0, 7] | 4.09e-2 / 9.21e-4 |
| 200 | 111.27972 / 109.62785 | 9.10 | 8.36 | [0] / [2, 4, 5, 14] | 1.547e-1 / 3.19e-3 |

That is tracked in gam:
- gam#2954: the stationarity band grows with the row count, so the default fit certifies at |g| ≈ 1e-3·(1+n).
- gam#2953: a certified point gets published while a refused iterate has a lower criterion. At n=200 one seed reached 109.9206, but 111.2797 was published.

This issue stays open until those fixes reach a pinned gam release.

**Comment by SauersML** [comment]
**Multistart landed in gam** (strike-lanes, for the #2359 strike team)

- SauersML/gam@8b19e9d367: every generated BMS outer seed searches in its own run, in parallel, and the fit publishes the best certified one. This is gnomon-2370's multistart v4 plus per-seed verdicts in the refusal.
- SauersML/gam@2af589fae8: each search gets its own same-β exact-cache and rigid-tensor stores. Members had shared the process-wide store, so whether a member hit the parent's β0 entry or rebuilt it (which depended on eviction order) decided its trajectory. This commit also adds a pin: a member's values are bit-identical at one lane and at six. The pin fails when members share the global store.

**Lane gate** (n200 = the 200-row calibration). Every seed's value, final ρ, iteration count and verdict, and the saved model, are bitwise identical across:
- 6 and 1 lanes on 12 threads, and 4 lanes on 4 threads, gated on gam a339bedd1b;
- 6, 3 and t4 on 818e73cfbf;
- 6 (twice), 4, 2, 1 and t4 on e9c0e0028f, for both the n200 and 1-PC fixtures.

| seed | start ρ | certified value | iterations |
|---|---|---|---|
| 0 | 0 | 128.27970789781477 | 17 |
| 1 | 1 | **128.27907526189674** (winner) | 14 |
| 2 | [22.78, 20.93, …] | 128.32317526887812 | 0 |
| 3 | 2 | 128.32350684044954 | 6 |
| 4 | 4 | 128.32605933974892 | 6 |
| 5 | −2 | 128.32414818318395 | 7 |

**gam tests** on a339bedd1b plus both commits all pass: gam-solve lib 1961, gam-custom-family lib 336, gam-models `bms::` 294, and the member and multistart pins. The ban-scanner workspace build passes. The drift check exits 0.

**Memory at n200.** (RSS at 6 lanes − RSS at 1 lane)/5 is 13.1–16.0 MiB, against the 12.5 MiB charged per search. A live-heap tracker shows this gap is not per search:
- The six searches in flight hold ≤ 1.5 MiB each beyond the launch heap.
- The gap is one retained 16 MiB (2^24 B) scratch chunk per pool worker thread. There are 12 chunks (the pool width) during and after the multistart at 6 lanes, and a serial run accumulates them over time.
- At a fixed 4 lanes, peak RSS is 121,916 KiB on 4 threads and 243,532 KiB on 12.

The lower-basin question is unchanged by this landing. At stock tolerance the winner is seed 1 at 128.2791, and reaching the lower basin still waits on stage 1b.

**Comment by SauersML** [comment]
**Single-start+1b vs multistart+1b A/B** (strike-lanes, for the #2359 multistart decision). **Multistart+1b certifies the lower V on the #2359 fixture, so multistart stays as the only behaviour.**

**Harness**
- gamfit on gam 466afd68b5, built as a test-profile cdylib, plus two lane-only patches:
  - strike-1b's stage-1b part 1 (`s1b_p1c.patch`, identical to its landing mbox);
  - an env switch that restores the pre-multistart single screened start (seed_budget 1, single cascade). The switch never goes in a landing tree.
- The n=200 fit reproduces gnomon calibrate's `--num-pcs 0` BMS spec. Validated without 1b: single start 128.27970789743654 and multistart winner 128.27907526117318 match gnomon's own n200 values (128.279707898 and 128.279075262), with the same ρ.
- Every certificate below is on the literal **solver-band** bound (4.020e-3 at n=200, 4.002e-2 at K=16). strike-1b's part 2 replaces it with the derived band, so re-read these verdicts then. In particular, re-read seed 4's non-certification on n=200.

**#2359 n=200 fixture** (12 threads, acn112)

| arm | seed | start ρ | V | verdict | iterations | seed wall |
|---|---|---|---|---|---|---|
| single+1b | screened start | 0 | 128.27970789769654 | \|Pg\|=2.066e-3 ≤ 4.020e-3, stationary; 1b's mint Hessian withdrew λ_min=−1.29e-6 (criterion-contradicted) | — | 252 s total, 683 vCPU-s |
| multistart+1b | 0 | 0 | 128.27970789744612 | certified | 17 | 677 s |
| | 1 | 1 | 128.27907526117318 | certified | 14 | 667 s |
| | 2 | upper rails | 128.3231751 | certified | 0 | — |
| | 3 | 2 | 128.2767348742336 | certified after a 1b escape | 29 | 2,257 s |
| | 4 | 4 | not certified | \|g\|=2.185e-2 > 4.020e-3, λ_min(H)=−6.648e-3 | — | 2,140 s |
| | **5** | **−2** | **126.65965524058215** | **\|g\|=3.609e-3 ≤ 4.020e-3, hessian_psd=yes (terminal analytic), stationary** | 22 | 1,612 s |
| multistart+1b total | winner seed 5 | | **126.6597**, 1.620 nats below single+1b | | | 2,259 s wall, 21,568 vCPU-s (31.6×) |

- Winner ρ: [13.52, −6.17, −9.37, −6.12, 1.88, −0.78, 9.94, −6.76, −1.73].
- Single+1b ρ: [22.78, 4.59, 8.75, 10.51, 2.14, −1.45, −4.44, 0.14, −0.42].
- Without 1b, multistart wins by only 6.3e-4 nats, at 1,393 vs 339 vCPU-s.
- The seeds' spread rules out an adaptive seed count. Seeds 0–4 lie within 0.047 nats of one another (128.2767–128.3261), and only the last seed reaches 126.66.

**K=16 production BMS fixture** (study-speed-bin's `bin_k16`, the first 2,000 rows; 19 ρ, p=76; 16 threads, msismall)
- **With 1b:**
  - single+1b did not finish in 70 min (59,159 vCPU-s). The first-order check at 855.7034 passed (\|g\|=2.927e-2 ≤ 4.002e-2). 1b's mint exact Hessian then took about 25 min on this pre-e3f4bbedd9 base and confirmed a strict saddle (λ_min=−1.364e-3). The run was still escaping at the cap.
  - multistart+1b was not run: six searches at that Hessian cost cannot finish.
- **Without 1b:**

  | arm | V | seeds | vCPU-s | wall | RMSE vs p_true |
  |---|---|---|---|---|---|
  | single | 855.7033766095484, \|g\|=2.927e-2 ≤ 4.002e-2 | — | 7,090 | 490 s | 0.03486 |
  | multistart | **855.5639090200953** (seed 3) | 855.5639, 855.5708, 855.7034, 855.7053, 857.1783 (ρ=−2), 999.1153 (railed seed, 0 iterations) | 38,393 | 2,557 s | 0.04029 |

  - Eval-set Δrisk between the arms (20k rows): max \|Δrisk\| 0.136, which is 0.785 of the risk SD pooled; within strata eur 0.82, amr 0.69, afr 0.49, sas 0.42, eas 0.36.
  - At n=2,000 the lower-criterion basin is further from the truth.

Still to do: re-measure both fixtures on gam main with 1b and gam-2922's e3f4bbedd9 Hessian kernel built in.

**Comment by SauersML** [comment]
**[basin sensitivity]** A calibration's predictions depend on which certified basin multistart publishes, and at K=16, n=2,000 the lower-V basin predicts worse (strike-lanes)

This comes from the multistart A/B above: the production K=16 fit (study-speed-bin's `bin_k16`, first 2,000 rows, 19 ρ, p=76), without stage 1b, gam 466afd68b5.

- **Two certified basins, 0.139 nats apart:**
  - single start: 855.7033766, |g| 2.927e-2 ≤ 4.002e-2 on the literal solver band;
  - multistart winner: seed 3, 855.5639090.
  - The other seeds certify at 855.5708, 855.7053 and 857.1783, and the railed seed at 999.1153.
- **Their predictions differ by most of a risk SD.** On the 20,000 eval rows, max |Δrisk| is 0.136, 0.785 of the risk SD pooled. Within strata: eur 0.82, amr 0.69, afr 0.49, sas 0.42, eas 0.36.
- **Lower criterion is not closer to the truth here.** RMSE against p_true is 0.04029 for the multistart winner and 0.03486 for the single start, 16% worse for the lower-V basin.

**What this means for gnomon.** The published model is the argmin of the fit criterion over certified basins, and that is the rule multistart implements. But two basins within 0.14 nats of each other (well inside any practical model-selection resolution) move individual risks by up to 0.79 SD. So at this n the calibrated risks are not determined by the data to that precision. The basin spread is a property of the fit, not a solver defect.

**Not yet shown:** whether this persists at n=10k, which is AoU-like. One evaluation takes about 50 s there, and 1b's mint Hessian on this pre-e3f4bbedd9 base took about 25 min at n=2,000. The n=10k A/B is owed on a build with gam-2922's kernel.

**Comment by SauersML** [comment]
**[mint exact Hessian on an all-railed face: parked]** (strike-1b)

Proposal (i): skip the mint's order-four Hessian when every coordinate is railed and the projected gradient is identically zero. On an empty free face nothing reads H: the decrement verdict, the rail certificate, the flatness rung and the V_rho correction all see empty sub-blocks.

**Measured.** On gnomon-2370's t64 BMS binary fixture, with the #2405 secant route and no mid-run reseed, every Mint certificate is interior. At most 1 of 9 coordinates is railed (rho_0 at 21.49 or 19.27), and those mints certify on curvature-resolvability or coordinate-band. So the skip buys 0 exact Hessians there. The one all-railed mint I have is the earlier p1c probe of this issue's 50-row fixture (railed=[0..8], |Pg|=0, |g|=2.9e-9).

**Cost of doing it.** Published hessian_psd would go from true to null on those fits. It also needs a vacuous-PSD curvature evidence variant, and first_order_smoothing_correction taking Option<H>.

**Decision.** Parked. It gets built only when a measured production fit mints at an all-railed corner and pays real time for that Hessian.
